src/NzbDrone.Common/Disk/DiskTransferService.cs

```diff
@@ -0,0 +1,267 @@
+
+            if (sourcePath.PathEquals(targetPath))
+            {
+                _logger.Warn("Source and destination can't be the same {0}", sourcePath);
+                return TransferMode.None;
+            }
+
+            if (_diskProvider.FileExists(targetPath) && overwrite)
+            {
+                _diskProvider.DeleteFile(targetPath);
+            }
+
+            if (mode.HasFlag(TransferMode.HardLink))
+            {
+                bool createdHardlink = _diskProvider.TryCreateHardLink(sourcePath, targetPath);
+                if (createdHardlink)
+                {
+                    return TransferMode.HardLink;
+                }
+                if (!mode.HasFlag(TransferMode.Copy))
+                {
+                    throw new IOException("Hardlinking from '" + sourcePath + "' to '" + targetPath + "' failed.");
+                }
+            }
+
+            if (mode.HasFlag(TransferMode.Copy))
+            {
+                _diskProvider.CopySingleFile(sourcePath, targetPath);
+                return TransferMode.Copy;
+            }
+
+            if (mode.HasFlag(TransferMode.Move))
+            {
+                _diskProvider.MoveSingleFile(sourcePath, targetPath);
+                return TransferMode.Move;
+            }
+
+            return TransferMode.None;
+        }
+
+        public TransferMode TransferFileVerified(String sourcePath, String targetPath, TransferMode mode)
+        {
+            Ensure.That(sourcePath, () => sourcePath).IsValidPath();
+            Ensure.That(targetPath, () => targetPath).IsValidPath();
+
+            _logger.Debug("{0}Verified [{1}] > [{2}]", mode, sourcePath, targetPath);
+
+            if (sourcePath.PathEquals(targetPath))
+            {
+                _logger.Warn("Source and destination can't be the same {0}", sourcePath);
+                return TransferMode.None;
+            }
+
+            if (mode.HasFlag(TransferMode.HardLink))
+            {
+                if (_diskProvider.TryCreateHardLink(sourcePath, targetPath))
+                {
+                    return TransferMode.HardLink;
+                }
+            }
+
+            if (mode.HasFlag(TransferMode.Copy))
+            {
+                if (TryCopyFile(sourcePath, targetPath))
+                {
+                    return TransferMode.Copy;
+                }
+            }
+
+            if (mode.HasFlag(TransferMode.Move))
+            {
+                if (TryMoveFile(sourcePath, targetPath))
+                {
+                    return TransferMode.Move;
+                }
+            }
+
+            throw new IOException(String.Format("Failed to completely transfer [{0}] to [{1}], aborting.", sourcePath, targetPath));
+        }
+
+        private Boolean TryCopyFile(String sourcePath, String targetPath)
+        {
+            var originalSize = _diskProvider.GetFileSize(sourcePath);
+
+            for (var i = 0; i <= RetryCount; i++)
```

Comment:
would this ever pass when tried right after? shouldn't we put some sleep in there?